File: LocalServer/libs/hubs_api.py
import logging
import jsonpickle
import threading
from wshubsapi import utils
from concurrent.futures import Future
import asyncio

utils.set_serializer_date_handler()
_message_id = 0
_message_lock = threading.RLock()


class GenericClient(object):
    def __setattr__(self, key, value):
        return super(GenericClient, self).__setattr__(key, value)


class GenericServer(object):
    def __init__(self, hub):
        self.hub = hub

    @classmethod
    def _get_next_message_id(cls):
        global _message_id
        with _message_lock:
            _message_id += 1
            return _message_id

    def _serialize_object(self, obj2ser):
        return jsonpickle.encode(obj2ser, **self.hub.serialization_args)

    def construct_message(self, args, function_name):
        id_ = self._get_next_message_id()
        body = {"hub": self.hub.name, "function": function_name, "args": args, "ID": id_}
        future = self.hub.ws_client.get_future(id_)
        send_return_obj = self.hub.ws_client.send(self._serialize_object(body))
        if isinstance(send_return_obj, Future):
            return send_return_obj
        else:
            return future


class GenericBridge(GenericServer):
    def __getattr__(self, function_name):
        def function_wrapper(*args_array):
            """
            :rtype : Future
            """
            args = list()
            args.append(self.clients_ids)
            args.append(function_name)
            args.append(args_array)
            id_ = self._get_next_message_id()
            body = {"hub": self.hub.name, "function": "_client_to_clients_bridge", "args": args, "ID": id_}
            future = self.hub.ws_client.get_future(id_)
            send_return_obj = self.hub.ws_client.send(self._serialize_object(body))
            if isinstance(send_return_obj, Future):
                return send_return_obj
            return future

        return function_wrapper


def construct_api_client_class(client_class):
    if client_class is None:
        from ws4py.client.threadedclient import WebSocketClient
        client_class = WebSocketClient

    class WSHubsAPIClient(client_class):
        def __init__(self, api, url):
            """
            :type api: HubsAPI
            """
            client_class.__init__(self, url)
            self.__futures = dict()
            self.is_opened = False
            self.api = api
            self.log = logging.getLogger(__name__)
            self.log.addHandler(logging.NullHandler())

        def opened(self):
            self.is_opened = True
            self.log.debug("Connection opened")

        def closed(self, code, reason=None):
            self.log.debug("Connection closed with code:\n%s\nAnd reason:\n%s" % (code, reason))

        def received_message(self, m):
            try:
                msg_obj = jsonpickle.decode(m.data.decode('utf-8'))
            except Exception as e:
                self.on_error(e)
                return
            if "reply" in msg_obj:
                f = self.__futures.get(msg_obj["ID"], None)
                if f is None:
                    return
                if msg_obj["success"]:
                    f.set_result(msg_obj["reply"])
                else:
                    f.set_exception(Exception(msg_obj["reply"]))
            else:
                try:
                    client_function = getattr(getattr(self.api, (msg_obj["hub"])).client, msg_obj["function"])
                    replay_message = dict(ID=msg_obj["ID"])
                    try:
                        reply = client_function(*msg_obj["args"])
                        replay_message["reply"] = reply
                        replay_message["success"] = True
                    except Exception as e:
                        replay_message["reply"] = str(e)
                        replay_message["success"] = False
                    finally:
                        self.api.ws_client.send(self.api.serialize_object(replay_message))
                except:
                    self.log.exception("unable to call client function")

            self.log.debug("Message received: %s" % m.data.decode('utf-8'))

        def get_future(self, id_):
            """
            :rtype : Future
            """
            self.__futures[id_] = Future()
            return self.__futures[id_]

        def on_error(self, exception):
            self.log.exception("Error in protocol")

        def default_on_error(self, error):
            pass

    return WSHubsAPIClient


class HubsAPI(object):
    def __init__(self, url, client_class=None, serialization_max_depth=5, serialization_max_iter=100):
        api_client_class = construct_api_client_class(client_class)
        self.ws_client = api_client_class(self, url)
        self.ws_client.default_on_error = lambda error: None
        self.serialization_args = dict(max_depth=serialization_max_depth, max_iter=serialization_max_iter)
        self.serialization_args['unpicklable'] = True
        self.UserHub = self.UserHubClass(self.ws_client, self.serialization_args)
        self.HouseHub = self.HouseHubClass(self.ws_client, self.serialization_args)
        self.UtilsAPIHub = self.UtilsAPIHubClass(self.ws_client, self.serialization_args)

    @property
    def default_on_error(self):
        return None

    @default_on_error.setter
    def default_on_error(self, func):
        self.ws_client.default_on_error = func

    def connect(self):
        self.ws_client.connect()

    def serialize_object(self, obj2ser):
        return jsonpickle.encode(obj2ser, self.serialization_args)

    class UserHubClass(object):
        def __init__(self, ws_client, serialization_args):
            self.name = "UserHub"
            self.ws_client = ws_client
            self.serialization_args = serialization_args
            self.server = self.ServerClass(self)
            self.client = self.ClientClass()

        def get_clients(self, client_ids):
            return HubsAPI.UserHubClass.ClientsInServer(client_ids, self)

        class ServerClass(GenericServer):
            
            def register(self, user_json):
                """
                :rtype : Future
                """
                args = list()
                args.append(user_json)
                return self.construct_message(args, "register")

            def get_my_houses(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "get_my_houses")

            def add_house(self, house):
                """
                :rtype : Future
                """
                args = list()
                args.append(house)
                return self.construct_message(args, "add_house")

            def subscribe_to_hub(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "subscribe_to_hub")

            def unsubscribe_from_hub(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "unsubscribe_from_hub")

            def login(self, user_json):
                """
                :rtype : Future
                """
                args = list()
                args.append(user_json)
                return self.construct_message(args, "login")

            def get_subscribed_clients_ids(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "get_subscribed_clients_ids")

            def remove_house(self, house_id):
                """
                :rtype : Future
                """
                args = list()
                args.append(house_id)
                return self.construct_message(args, "remove_house")

        class ClientClass(GenericClient):
            def __init__(self):
                pass
            

        class ClientsInServer(GenericBridge):
            def __init__(self, client_ids, hub):
                super(self.__class__, self).__init__(hub)
                self.clients_ids = client_ids
            

    class HouseHubClass(object):
        def __init__(self, ws_client, serialization_args):
            self.name = "HouseHub"
            self.ws_client = ws_client
            self.serialization_args = serialization_args
            self.server = self.ServerClass(self)
            self.client = self.ClientClass()

        def get_clients(self, client_ids):
            return HubsAPI.HouseHubClass.ClientsInServer(client_ids, self)

        class ServerClass(GenericServer):
            
            def create(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "create")

            def get_instance(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "get_instance")

            def subscribe_to_hub(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "subscribe_to_hub")

            def list_houses(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "list_houses")

            def unsubscribe_from_hub(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "unsubscribe_from_hub")

            def get_subscribed_clients_ids(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "get_subscribed_clients_ids")

        class ClientClass(GenericClient):
            def __init__(self):
                pass
            
            def reset_module(self, module_id):
                pass

            def component_read(self, module_id, component_index):
                pass

            def stop_module_communication(self, module_id):
                pass

            def component_write(self, module_id, component_index, value):
                pass

            def get_components(self, ):
                pass

        class ClientsInServer(GenericBridge):
            def __init__(self, client_ids, hub):
                super(self.__class__, self).__init__(hub)
                self.clients_ids = client_ids
            
            def reset_module(self, module_id):
                """
                :rtype : Future
                """
                args = list()
                args.append(self.clients_ids)
                args.append("reset_module")
                args.append([module_id])
                return self.construct_message(args, "_client_to_clients_bridge") 

            def component_read(self, module_id, component_index):
                """
                :rtype : Future
                """
                args = list()
                args.append(self.clients_ids)
                args.append("component_read")
                args.append([module_id, component_index])
                return self.construct_message(args, "_client_to_clients_bridge") 

            def stop_module_communication(self, module_id):
                """
                :rtype : Future
                """
                args = list()
                args.append(self.clients_ids)
                args.append("stop_module_communication")
                args.append([module_id])
                return self.construct_message(args, "_client_to_clients_bridge") 

            def component_write(self, module_id, component_index, value):
                """
                :rtype : Future
                """
                args = list()
                args.append(self.clients_ids)
                args.append("component_write")
                args.append([module_id, component_index, value])
                return self.construct_message(args, "_client_to_clients_bridge") 

            def get_components(self, ):
                """
                :rtype : Future
                """
                args = list()
                args.append(self.clients_ids)
                args.append("get_components")
                args.append([])
                return self.construct_message(args, "_client_to_clients_bridge") 

    class UtilsAPIHubClass(object):
        def __init__(self, ws_client, serialization_args):
            self.name = "UtilsAPIHub"
            self.ws_client = ws_client
            self.serialization_args = serialization_args
            self.server = self.ServerClass(self)
            self.client = self.ClientClass()

        def get_clients(self, client_ids):
            return HubsAPI.UtilsAPIHubClass.ClientsInServer(client_ids, self)

        class ServerClass(GenericServer):
            
            def subscribe_to_hub(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "subscribe_to_hub")

            def get_hubs_structure(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "get_hubs_structure")

            def set_id(self, client_id):
                """
                :rtype : Future
                """
                args = list()
                args.append(client_id)
                return self.construct_message(args, "set_id")

            def get_id(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "get_id")

            def unsubscribe_from_hub(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "unsubscribe_from_hub")

            def is_client_connected(self, client_id):
                """
                :rtype : Future
                """
                args = list()
                args.append(client_id)
                return self.construct_message(args, "is_client_connected")

            def get_subscribed_clients_ids(self, ):
                """
                :rtype : Future
                """
                args = list()
                
                return self.construct_message(args, "get_subscribed_clients_ids")

        class ClientClass(GenericClient):
            def __init__(self):
                pass
            

        class ClientsInServer(GenericBridge):
            def __init__(self, client_ids, hub):
                super(self.__class__, self).__init__(hub)
                self.clients_ids = client_ids
            

File: HousticApp/app/libs/hubsApi.js
'use strict';
/* jshint ignore:start */
/* ignore jslint start */
function HubsAPI(serverTimeout, wsClientClass, PromiseClass) {

    var messageID = 0,
        promisesHandler = {},
        defaultRespondTimeout = serverTimeout || 5000,
        thisApi = this,
        messagesBeforeOpen = [],
        emptyFunction = function () {},
        onOpenTriggers = [];

    PromiseClass = PromiseClass || Promise;
    if (!PromiseClass.prototype.finally) {
        PromiseClass.prototype.finally = function (callback) {
            var p = this.constructor;
            return this.then(
                function (value) {
                    return p.resolve(callback()).then(function () {
                        return value;
                    });
                },
                function (reason) {
                    return p.resolve(callback()).then(function () {
                        throw reason;
                    });
                });
        };
    }

    if (!PromiseClass.prototype.setTimeout) {
        PromiseClass.prototype.setTimeout = function (timeout) {
            clearTimeout(this._timeoutID);
            setTimeout(timeoutError(this._reject), timeout);
            return this;
        };
    }

    function timeoutError(reject) {
        return function () {
            reject(new Error('timeout error'));
        };
    }

    function toCamelCase(str) {
        return str.replace(/_([a-z])/g, function (g) { return g[1].toUpperCase(); });
    }

    this.clearTriggers = function () {
        messagesBeforeOpen = [];
        onOpenTriggers = [];
    };

    this.connect = function (url, reconnectTimeout) {
        return new PromiseClass(function (resolve, reject) {
            reconnectTimeout = reconnectTimeout || -1;
            function reconnect(error) {
                if (reconnectTimeout !== -1) {
                    window.setTimeout(function () {
                        thisApi.connect(url, reconnectTimeout);
                        thisApi.callbacks.onReconnecting(error);
                    }, reconnectTimeout * 1000);
                }
            }

            try {
                thisApi.wsClient = wsClientClass === undefined ? new WebSocket(url) : new wsClientClass(url);
            } catch (error) {
                reconnect(error);
                reject(error);
            }

            thisApi.wsClient.onopen = function () {
                resolve();
                thisApi.callbacks.onOpen(thisApi);
                onOpenTriggers.forEach(function (trigger) {
                    trigger();
                });
                messagesBeforeOpen.forEach(function (message) {
                    thisApi.wsClient.send(message);
                });
            };

            thisApi.wsClient.onclose = function (error) {
                reject(error);
                thisApi.callbacks.onClose(error);
                reconnect(error);
            };

            thisApi.wsClient.addOnOpenTrigger = function (trigger) {
                if (thisApi.wsClient.readyState === 0) {
                    onOpenTriggers.push(trigger);
                } else if (thisApi.wsClient.readyState === 1) {
                    trigger();
                } else {
                    throw new Error('web socket is closed');
                }
            };

            thisApi.wsClient.onmessage = function (ev) {
                try {
                    var promiseHandler,
                        msgObj = JSON.parse(ev.data);
                    if (msgObj.hasOwnProperty('reply')) {
                        promiseHandler = promisesHandler[msgObj.ID];
                        msgObj.success ? promiseHandler.resolve(msgObj.reply) : promiseHandler.reject(msgObj.reply);
                    } else {
                        msgObj.function = toCamelCase(msgObj.function);
                        var executor = thisApi[msgObj.hub].client[msgObj.function];
                        if (executor !== undefined) {
                            var replayMessage = {ID: msgObj.ID};
                            try {
                                replayMessage.reply = executor.apply(executor, msgObj.args);
                                replayMessage.success = true;
                            } catch (e) {
                                replayMessage.success = false;
                                replayMessage.reply = e.toString();
                            } finally {
                                if (replayMessage.reply instanceof PromiseClass) {
                                    replayMessage.reply.then(function (result) {
                                        replayMessage.success = true;
                                        replayMessage.reply = result;
                                        thisApi.wsClient.send(JSON.stringify(replayMessage));
                                    }, function (error) {
                                        replayMessage.success = false;
                                        replayMessage.reply = error;
                                        thisApi.wsClient.send(JSON.stringify(replayMessage));
                                    });
                                } else {
                                    replayMessage.reply = replayMessage.reply === undefined ? null : replayMessage.reply;
                                    thisApi.wsClient.send(JSON.stringify(replayMessage));
                                }
                            }
                        } else {
                            thisApi.callbacks.onClientFunctionNotFound(msgObj.hub, msgObj.function);
                        }
                    }
                } catch (err) {
                    thisApi.wsClient.onMessageError(err);
                }
            };

            thisApi.wsClient.onMessageError = function (error) {
                thisApi.callbacks.onMessageError(error);
            };
        });
    };

    this.callbacks = {
        onClose: emptyFunction,
        onOpen: emptyFunction,
        onReconnecting: emptyFunction,
        onMessageError: emptyFunction,
        onClientFunctionNotFound: emptyFunction
    };

    this.defaultErrorHandler = null;

    var constructMessage = function (hubName, functionName, args) {
        if (thisApi.wsClient === undefined) {
            throw new Error('ws not connected');
        }
        var promise,
            timeoutID = null,
            _reject;
        promise = new PromiseClass(function (resolve, reject) {
            args = Array.prototype.slice.call(args);
            var id = messageID++,
                body = {'hub': hubName, 'function': functionName, 'args': args, 'ID': id};
            promisesHandler[id] = {};
            promisesHandler[id].resolve = resolve;
            promisesHandler[id].reject = reject;
            timeoutID = setTimeout(timeoutError(reject), defaultRespondTimeout);
            _reject = reject;

            if (thisApi.wsClient.readyState === WebSocket.CONNECTING) {
                messagesBeforeOpen.push(JSON.stringify(body));
            } else if (thisApi.wsClient.readyState !== WebSocket.OPEN) {
                reject('webSocket not connected');
            } else {
                thisApi.wsClient.send(JSON.stringify(body));
            }
        });
        promise._timeoutID = timeoutID;
        promise._reject = _reject;
        return promise;
    };
    
    this.HouseHub = {};
    this.HouseHub.server = {
        __HUB_NAME : 'HouseHub',
        
        subscribeToHub : function (){
            
            return constructMessage('HouseHub', 'subscribe_to_hub', arguments);
        },

        listHouses : function (){
            
            return constructMessage('HouseHub', 'list_houses', arguments);
        },

        unsubscribeFromHub : function (){
            
            return constructMessage('HouseHub', 'unsubscribe_from_hub', arguments);
        },

        getAllComponents : function (houseId){
            
            return constructMessage('HouseHub', 'get_all_components', arguments);
        },

        getInstance : function (){
            
            return constructMessage('HouseHub', 'get_instance', arguments);
        },

        resetModule : function (houseId, moduleId){
            
            return constructMessage('HouseHub', 'reset_module', arguments);
        },

        getSubscribedClientsIds : function (){
            
            return constructMessage('HouseHub', 'get_subscribed_clients_ids', arguments);
        },

        componentWrite : function (houseId, moduleId, componentKey, value){
            
            return constructMessage('HouseHub', 'component_write', arguments);
        },

        stopModuleCommunication : function (houseId, moduleId){
            
            return constructMessage('HouseHub', 'stop_module_communication', arguments);
        },

        componentRead : function (houseId, moduleId, componentKey){
            
            return constructMessage('HouseHub', 'component_read', arguments);
        },

        create : function (){
            
            return constructMessage('HouseHub', 'create', arguments);
        }
    };
    this.HouseHub.client = {};
    this.UserHub = {};
    this.UserHub.server = {
        __HUB_NAME : 'UserHub',
        
        subscribeToHub : function (){
            
            return constructMessage('UserHub', 'subscribe_to_hub', arguments);
        },

        unsubscribeFromHub : function (){
            
            return constructMessage('UserHub', 'unsubscribe_from_hub', arguments);
        },

        login : function (userJson){
            
            return constructMessage('UserHub', 'login', arguments);
        },

        getSubscribedClientsIds : function (){
            
            return constructMessage('UserHub', 'get_subscribed_clients_ids', arguments);
        },

        getMyHouses : function (){
            
            return constructMessage('UserHub', 'get_my_houses', arguments);
        },

        removeHouse : function (houseId){
            
            return constructMessage('UserHub', 'remove_house', arguments);
        },

        register : function (userJson){
            
            return constructMessage('UserHub', 'register', arguments);
        },

        addHouse : function (house){
            
            return constructMessage('UserHub', 'add_house', arguments);
        }
    };
    this.UserHub.client = {};
    this.UtilsAPIHub = {};
    this.UtilsAPIHub.server = {
        __HUB_NAME : 'UtilsAPIHub',
        
        subscribeToHub : function (){
            
            return constructMessage('UtilsAPIHub', 'subscribe_to_hub', arguments);
        },

        getHubsStructure : function (){
            
            return constructMessage('UtilsAPIHub', 'get_hubs_structure', arguments);
        },

        getId : function (){
            
            return constructMessage('UtilsAPIHub', 'get_id', arguments);
        },

        getSubscribedClientsIds : function (){
            
            return constructMessage('UtilsAPIHub', 'get_subscribed_clients_ids', arguments);
        },

        setId : function (clientId){
            
            return constructMessage('UtilsAPIHub', 'set_id', arguments);
        },

        isClientConnected : function (clientId){
            
            return constructMessage('UtilsAPIHub', 'is_client_connected', arguments);
        },

        unsubscribeFromHub : function (){
            
            return constructMessage('UtilsAPIHub', 'unsubscribe_from_hub', arguments);
        }
    };
    this.UtilsAPIHub.client = {};
}
/* jshint ignore:end */
/* ignore jslint end */
    